src/network/networkd-mdb.c

```diff
@@ -0,0 +1,252 @@
+        assert(lvalue);
+        assert(rvalue);
+        assert(data);
+
+        r = mdb_entry_new_static(network, filename, section_line, &mdb_entry);
+        if (r < 0)
+                return log_oom();
+
+        r = in_addr_from_string_auto(rvalue, &mdb_entry->family, &mdb_entry->group_addr);
+        if (r < 0) {
+                log_syntax(unit, LOG_WARNING, filename, line, r, "Cannot parse group address: %m");
+                return 0;
+        }
+
+        if (!in_addr_is_multicast(mdb_entry->family, &mdb_entry->group_addr)) {
+                log_syntax(unit, LOG_WARNING, filename, line, 0, "Address %s is not a multicast group", rvalue);
+                return 0;
+        }
+
+        mdb_entry = NULL;
+
+        return 0;
+}
+
+/* remove and MDB entry. */
+MdbEntry *mdb_entry_free(MdbEntry *mdb_entry) {
+        if (!mdb_entry)
+                return NULL;
+
+        if (mdb_entry->network) {
+                LIST_REMOVE(static_mdb_entries, mdb_entry->network->static_mdb_entries, mdb_entry);
+                assert(mdb_entry->network->n_static_mdb_entries > 0);
+                mdb_entry->network->n_static_mdb_entries--;
+
+                if (mdb_entry->section)
+                        hashmap_remove(mdb_entry->network->mdb_entries_by_section, mdb_entry->section);
+        }
+
+        network_config_section_free(mdb_entry->section);
+        free(mdb_entry);
+
+        return NULL;
+}
+
+static int set_mdb_handler(sd_netlink *rtnl, sd_netlink_message *m, Link *link) {
+        int r;
+
+        assert(link);
+
+        if (IN_SET(link->state, LINK_STATE_FAILED, LINK_STATE_LINGER))
+                return 1;
+
+        r = sd_netlink_message_get_errno(m);
+        if (r < 0 && r != -EEXIST) {
+                log_link_message_warning_errno(link, m, r, "Could not add MDB entry");
+                link_enter_failed(link);
+                return 1;
+        }
+
+        return 1;
+}
+
+/* send a request to the kernel to add an MDB entry */
+int mdb_entry_configure(Link *link, MdbEntry *mdb_entry) {
+        _cleanup_(sd_netlink_message_unrefp) sd_netlink_message *req = NULL;
+        struct br_mdb_entry entry = {};
+        int r;
+
+        assert(link);
+        assert(link->network);
+        assert(link->network->bridge);
+        assert(link->manager);
+        assert(mdb_entry);
+
+        entry.state = MDB_PERMANENT;
+        entry.ifindex = link->ifindex;
+        entry.vid = mdb_entry->vlan_id;
+
+        /* create new RTM message */
+        r = sd_rtnl_message_new_mdb(link->manager->rtnl, &req, RTM_NEWMDB, link->network->bridge->ifindex);
+        if (r < 0)
+                return log_link_error_errno(link, r, "Could not create RTM_NEWMDB message: %m");
+
+        switch (mdb_entry->family) {
+        case AF_INET:
+                entry = (struct br_mdb_entry) {
+                        .addr.u.ip4 = mdb_entry->group_addr.in.s_addr,
+                        .addr.proto = htobe16(ETH_P_IP),
+                };
```

Comment:
this will override the three initializations above. I m a bit confused, where those there already in the older version of the PR? Either they have to be copied into into this structured initialization, or this cannot actually use structured initialization, but right now you override what you initialized above.

Reply:
Yes, exactly, that makes no sense, and yes, the members above were also set in the initial version. I reverted to the original approach.